controllers/agents/app/serviceclaim_controller.go

```diff
@@ -43,12 +53,146 @@ type ServiceClaimReconciler struct {
-	_ = log.FromContext(ctx)
+	l := log.FromContext(ctx).WithValues("namespace", req.Namespace, "name", req.Name)
+	l.Info("Reconciling service claim")
+
+	var sclaim primazaiov1alpha1.ServiceClaim
+	if err := r.Get(ctx, req.NamespacedName, &sclaim); err != nil {
+		l.Error(err, "Failed to retrieve ServiceClaim")
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	config, remote_namespace, err := workercluster.GetPrimazaKubeconfig(ctx, sclaim.Namespace, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	l.Info("remote cluster", "address", config.Host)
+
+	remote_client, err := client.New(config, client.Options{
+		Scheme: r.Client.Scheme(),
+		Mapper: r.Mapper,
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	objKey := client.ObjectKey{
+		Name:      constants.ApplicationAgentDeploymentName,
+		Namespace: req.NamespacedName.Namespace,
+	}
+	var deployment appsv1.Deployment
+	err = r.Get(ctx, objKey, &deployment)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// it should never happen that this controller does not find itself
+			l.Error(err,
+				"service agent deployment not found, that should be a bug",
+				"expected deployment name", constants.ServiceAgentDeploymentName)
+			panic(err)
+		}
+		return ctrl.Result{}, err
+	}
+
+	// examine DeletionTimestamp to determine if object is under deletion
+	if sclaim.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !controllerutil.ContainsFinalizer(&sclaim, ServiceClaimFinalizer) {
+			controllerutil.AddFinalizer(&sclaim, ServiceClaimFinalizer)
+			if err := r.Update(ctx, &sclaim); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if controllerutil.ContainsFinalizer(&sclaim, ServiceClaimFinalizer) {
+			// our finalizer is present, so lets handle any external dependency
+			sclaimCopy := r.createServiceClaimCopy(sclaim, deployment, remote_namespace)
+			if err := r.deleteExternalResources(ctx, sclaimCopy, remote_client); err != nil {
+				// if fail to delete the external dependency here, return with error
+				// so that it can be retried
+				return ctrl.Result{}, err
+			}
+
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(&sclaim, ServiceClaimFinalizer)
+			if err := r.Update(ctx, &sclaim); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
 
-	// TODO(user): your logic here
+	sclaimCopy := r.createServiceClaimCopy(sclaim, deployment, remote_namespace)
+	spec := sclaimCopy.Spec
+	op, err := controllerutil.CreateOrUpdate(ctx, remote_client, sclaimCopy, func() error {
+		sclaimCopy.Spec = spec
+		return nil
+	})
+	if err != nil {
```

Comment:
If we are getting an error on service claim create or update , then we should return the error here and not proceed with the status update on L146

Reply:
Fixed.